File: OpenHarmony/developtools/hapsigner/hapsigntool_cpp/codesigning/datastructure/src/merkle_tree_extension.cpp
/*
 * Copyright (c) 2024-2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "merkle_tree_extension.h"

namespace OHOS {
namespace SignatureTools {

MerkleTreeExtension::MerkleTreeExtension()
    : Extension(MERKLE_TREE_INLINED, MERKLE_TREE_EXTENSION_DATA_SIZE)
{
    merkleTreeSize = 0;
    merkleTreeOffset = 0;
}

MerkleTreeExtension::MerkleTreeExtension(int64_t merkleTreeSize, int64_t merkleTreeOffset,
    const std::vector<int8_t> rootHash) : Extension(MERKLE_TREE_INLINED, MERKLE_TREE_EXTENSION_DATA_SIZE)
{
    this->merkleTreeSize = merkleTreeSize;
    this->merkleTreeOffset = merkleTreeOffset;
    this->rootHash = rootHash;
}

MerkleTreeExtension::~MerkleTreeExtension()
{
}

int32_t MerkleTreeExtension::GetSize()
{
    return Extension::EXTENSION_HEADER_SIZE + MERKLE_TREE_EXTENSION_DATA_SIZE;
}

int64_t MerkleTreeExtension::GetMerkleTreeSize()
{
    return merkleTreeSize;
}

int64_t MerkleTreeExtension::GetMerkleTreeOffset()
{
    return merkleTreeOffset;
}

void MerkleTreeExtension::SetMerkleTreeOffset(int64_t offset)
{
    merkleTreeOffset = offset;
}

void MerkleTreeExtension::ToByteArray(std::vector<int8_t>& ret)
{
    std::unique_ptr<ByteBuffer> bf = std::make_unique<ByteBuffer>(ByteBuffer(
        Extension::EXTENSION_HEADER_SIZE + MERKLE_TREE_EXTENSION_DATA_SIZE));
    std::vector<int8_t> extByteArr;
    Extension::ToByteArray(extByteArr);
    bf->PutData(extByteArr.data(), extByteArr.size());
    bf->PutInt64(merkleTreeSize);
    bf->PutInt64(merkleTreeOffset);
    bf->PutData(rootHash.data(), rootHash.size());
    ret = std::vector<int8_t>(bf->GetBufferPtr(), bf->GetBufferPtr() + bf->GetCapacity());
}

MerkleTreeExtension* MerkleTreeExtension::FromByteArray(std::vector<int8_t>& bytes)
{
    std::unique_ptr<ByteBuffer> bf = std::make_unique<ByteBuffer>(ByteBuffer(bytes.size()));
    bf->PutData(bytes.data(), bytes.size());
    bf->Flip();
    int64_t inMerkleTreeSize = 0;
    bool flag = bf->GetInt64(inMerkleTreeSize);
    if (!flag || (inMerkleTreeSize % PAGE_SIZE_4K != 0)) {
        PrintErrorNumberMsg("VERIFY_ERROR", VERIFY_ERROR,
                            "The merkletree size must be an integer multiple of 4096");
        return nullptr;
    }
    int64_t inMerkleTreeOffset = 0;
    flag = bf->GetInt64(inMerkleTreeOffset);
    if (!flag || (inMerkleTreeOffset % PAGE_SIZE_4K != 0)) {
        PrintErrorNumberMsg("VERIFY_ERROR", VERIFY_ERROR,
                            "The merkletree offset must be an integer multiple of 4096");
        return nullptr;
    }
    std::vector<int8_t> inRootHash(ROOT_HASH_SIZE);
    bf->GetByte(inRootHash.data(), inRootHash.size());
    return new MerkleTreeExtension(inMerkleTreeSize, inMerkleTreeOffset, inRootHash);
}

}
}

File: OpenHarmony/developtools/hapsigner/hapsigntool/hap_sign_tool_lib/src/main/java/com/ohos/hapsigntool/codesigning/datastructure/MerkleTreeExtension.java
/*
 * Copyright (c) 2023-2023 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ohos.hapsigntool.codesigning.datastructure;

import com.ohos.hapsigntool.codesigning.exception.VerifyCodeSignException;
import com.ohos.hapsigntool.codesigning.utils.NumberUtils;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Arrays;
import java.util.Locale;

/**
 * Merkle tree extension is a type of Extension to store a merkle tree's information, i.e. size and root hash, ect.
 * <p>
 * structure
 * <p>
 * 1) u32 type
 * <p>
 * 2) u64 merkleTreeSize: the size of merkle tree
 * <p>
 * 3) u64 merkleTreeOffset: offset of the merkle tree by the start of the file.
 * <p>
 * 4) u8[64] rootHash: merkle tree root hash
 *
 * @since 2023/09/08
 */
public class MerkleTreeExtension extends Extension {
    /**
     * Type of MerkleTreeExtension
     */
    public static final int MERKLE_TREE_INLINED = 0x1;

    /**
     * Byte size of MerkleTreeExtension including merkleTreeSize, offset and root hash.
     */
    public static final int MERKLE_TREE_EXTENSION_DATA_SIZE = 80;

    private static final int ROOT_HASH_SIZE = 64;

    private final long merkleTreeSize;

    private long merkleTreeOffset;

    private byte[] rootHash;

    /**
     * Constructor for MerkleTreeExtension
     *
     * @param merkleTreeSize   Byte array representation of merkle tree
     * @param merkleTreeOffset merkle tree offset based on file start
     * @param rootHash         Root hash of the merkle tree
     */
    public MerkleTreeExtension(long merkleTreeSize, long merkleTreeOffset, byte[] rootHash) {
        super(MERKLE_TREE_INLINED, MERKLE_TREE_EXTENSION_DATA_SIZE);
        this.merkleTreeSize = merkleTreeSize;
        this.merkleTreeOffset = merkleTreeOffset;
        if (rootHash == null) {
            this.rootHash = new byte[ROOT_HASH_SIZE];
        } else {
            this.rootHash = Arrays.copyOf(rootHash, ROOT_HASH_SIZE);
        }
    }

    @Override
    public int size() {
        return Extension.EXTENSION_HEADER_SIZE + MERKLE_TREE_EXTENSION_DATA_SIZE;
    }

    public long getMerkleTreeSize() {
        return merkleTreeSize;
    }

    public long getMerkleTreeOffset() {
        return merkleTreeOffset;
    }

    public void setMerkleTreeOffset(long offset) {
        this.merkleTreeOffset = offset;
    }

    /**
     * Converts MerkleTreeExtension to a newly created byte array
     *
     * @return Byte array representation of MerkleTreeExtension
     */
    @Override
    public byte[] toByteArray() {
        ByteBuffer bf = ByteBuffer.allocate(size()).order(ByteOrder.LITTLE_ENDIAN);
        bf.put(super.toByteArray());
        bf.putLong(this.merkleTreeSize);
        bf.putLong(this.merkleTreeOffset);
        bf.put(this.rootHash);
        return bf.array();
    }

    /**
     * Init the MerkleTreeExtension by a byte array
     *
     * @param bytes Byte array representation of a MerkleTreeExtension object
     * @return a newly created MerkleTreeExtension object
     * @throws VerifyCodeSignException parsing result invalid
     */
    public static MerkleTreeExtension fromByteArray(byte[] bytes) throws VerifyCodeSignException {
        ByteBuffer bf = ByteBuffer.allocate(bytes.length).order(ByteOrder.LITTLE_ENDIAN);
        bf.put(bytes);
        bf.rewind();
        long inMerkleTreeSize = bf.getLong();
        if (!NumberUtils.isMultiple4K(inMerkleTreeSize)) {
            throw new VerifyCodeSignException("merkleTreeSize is not a multiple of 4096");
        }
        long inMerkleTreeOffset = bf.getLong();
        if (!NumberUtils.isMultiple4K(inMerkleTreeOffset)) {
            throw new VerifyCodeSignException("merkleTreeOffset is not a aligned to 4096");
        }
        byte[] inRootHash = new byte[ROOT_HASH_SIZE];
        bf.get(inRootHash);
        return new MerkleTreeExtension(inMerkleTreeSize, inMerkleTreeOffset, inRootHash);
    }

    /**
     * Return a string representation of the object
     *
     * @return string representation of the object
     */
    @Override
    public String toString() {
        return String.format(Locale.ROOT, "MerkleTreeExtension: merkleTreeSize[%d], merkleTreeOffset[%d],"
                + " rootHash[%s]", this.merkleTreeSize, this.merkleTreeOffset, Arrays.toString(this.rootHash));
    }
}
